File: kangaroo_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/userPinResetWithVerificationCode/UserPinResetHandler.kt
@file:Suppress("INLINE_FROM_HIGHER_PLATFORM")
@file:OptIn(ExperimentalJsExport::class)
package com.kangaroo.flutterplugin.android.features.userPinResetWithVerificationCode

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kotlin.js.ExperimentalJsExport
import features.userPinResetWithVerificationCode.UserPinResetApi

import kangaroorewards.appsdk.core.io.EmptyResponse
import features.userPinResetWithVerificationCode.serializeUserPinResetState
import kangaroorewards.appsdk.core.domain.SerializedResult
import kangaroorewards.appsdk.core.domain.toJsonResult


class UserPinResetHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/reset_pin"

    override val eventChannel: String
        get() = "customer_sdk/events/reset_pin"

    override suspend fun onMethodCall(call: MethodCall): String? {
        return resetPin(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        suspend fun resetPin(call: MethodCall): String? {
            val result = UserPinResetApi().resetPin(
                verificationCode = call.argument<String>("verificationCode") as String,
                pinCode = call.argument<String>("pinCode") as String,
                email = call.argument<String?>("email"),
                phone = call.argument<String?>("phone"),
                countryCode = call.argument<String?>("countryCode")
            ).toJsonResult<EmptyResponse>()

            return when (result) {
                is SerializedResult.Success -> result.data
                is SerializedResult.UnauthorizedError -> result.error
                is SerializedResult.UnknownError -> result.error
                else -> null
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        UserPinResetApi().userPinResetState.serializeUserPinResetState().observe { result ->
            pushSerializedResultToEventSink<UserPinResetApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}

File: kangaroo_app_sdk_ios/ios/Classes/features/userPinResetWithVerificationCode/UserPinResetHandler.swift
import Foundation
import Flutter
import KangarooAppSdkCustomer

class UserPinResetHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/reset_pin"

    var eventChannel: String = "customer_sdk/events/reset_pin"

    func onMethodCall(call: FlutterMethodCall) async -> Any? {
        return await UserPinResetHandler.resetPin(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func resetPin(call: FlutterMethodCall) async -> String? {
        

        


        

        guard let args = call.arguments else {
            return nil
        }
        do {
       if let myArgs = args as? [String: Any] {
                        guard let verificationCode = myArgs["verificationCode"] as? String else {return nil}
                guard let pinCode = myArgs["pinCode"] as? String else {return nil}
                let email = myArgs["email"] as? String? ?? nil
                let phone = myArgs["phone"] as? String? ?? nil
                let countryCode = myArgs["countryCode"] as? String? ?? nil

         let result = try await UserPinResetApi().resetPin(
                verificationCode: verificationCode,
                pinCode: pinCode,
                email: email,
                phone: phone,
                countryCode: countryCode
            ).serializeUserPinResetApiResult()

        switch result {
            case let result as SerializedResultSuccess:
                return result.data
            case let result as SerializedResultUnauthorizedError:
                return result.error
            case let result as SerializedResultUnknownError:
                return result.error
            default:
                return nil
                }
            }
        }
        catch {
            return nil
        }
        
        return nil
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        UserPinResetApi().userPinResetState
            .serializeUserPinResetState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultEmptyResponse:
                self.sink?(result.body)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
